Format method caching

frenzyigor opened the pull request:
added simple RandomizedLRU cache and interface ICUstomCache to use user defined format caching

Commit:
added simple RandomizedLRU cache and interface ICUstomCache to use user defined format caching

kochetkov commented on sources/LibProtection.Injections/Caching/ICustomCache.cs:
```diff
@@ -0,0 +1,8 @@
+    public interface ICustomCache
+    {
+        bool Add(CacheFormatItem formatItem, string result);
+        bool Get(CacheFormatItem formatItem, out string result);
```
This method should return both Success and Result cached values.

Commit:
added two FormatSuccess parameter to ICustomCache

Commit:
added Cache tuning method to SafeString interface

added ability to change RandomizedLRU cache size (or disable it if size equals zero)

Commit:
updated RandomizedLRUCache to lockfree realization

Commit:
changed RandomizedLRUCache interface back

Commit:
cached delegate for TryFormat

Commit:
added simple HashTable cache
